From 16cd15654cc66acc976dd6c4b4c5249561123d37 Mon Sep 17 00:00:00 2001
From: Eric Naim <dnaim@cachyos.org>
Date: Mon, 27 Jan 2020 18:21:09 +0100
Subject: [PATCH] cachy: micro-optimizations (curated backport)
Message-ID: <cachy-fork-cachyos-7.2-rc6-2-micro-opts>

Curated backport of CachyOS/linux fork micro-optimizations (cachyos-7.2-rc6-2):
  dec8b451162e mm: Increase VM_READAHEAD_PAGES to 256K
  1c37598e86ec sched/core: add branch hints based on gcov analysis
  4b167ad6a219 readdir: add unlikely hint on len check
  10fc94fd82d6 list: Mark some functions as __always_inline
  818b55203e93 Input: evdev - use call_rcu when detaching client

Signed-off-by: Eric Naim <dnaim@cachyos.org>
Signed-off-by: Colin Ian King <colin.i.king@gmail.com>
Signed-off-by: Kenny Levinsen <kl@kl.wtf>
---
diff --git a/include/linux/pagemap.h b/include/linux/pagemap.h
index 2c3718d592d6..42dba9e2da89 100644
--- a/include/linux/pagemap.h
+++ b/include/linux/pagemap.h
@@ -1363,7 +1363,7 @@ struct readahead_control {
 		._index = i,						\
 	}
 
-#define VM_READAHEAD_PAGES	(SZ_128K / PAGE_SIZE)
+#define VM_READAHEAD_PAGES	(SZ_256K / PAGE_SIZE)
 
 void page_cache_ra_unbounded(struct readahead_control *,
 		unsigned long nr_to_read, unsigned long lookahead_count);
-- 
2.55.0

diff --git a/kernel/sched/core.c b/kernel/sched/core.c
index 96226707c2f6..699491196a96 100644
--- a/kernel/sched/core.c
+++ b/kernel/sched/core.c
@@ -663,7 +663,7 @@ void raw_spin_rq_lock_nested(struct rq *rq, int subclass)
 
 	/* Matches synchronize_rcu() in __sched_core_enable() */
 	preempt_disable();
-	if (sched_core_disabled()) {
+	if (likely(sched_core_disabled())) {
 		raw_spin_lock_nested(&rq->__lock, subclass);
 		/* preempt_count *MUST* be > 1 */
 		preempt_enable_no_resched();
@@ -875,7 +875,7 @@ void update_rq_clock(struct rq *rq)
 	scx_rq_clock_update(rq, clock);
 
 	delta = clock - rq->clock;
-	if (delta < 0)
+	if (unlikely(delta < 0))
 		return;
 	rq->clock += delta;
 
@@ -6225,7 +6225,7 @@ pick_next_task(struct rq *rq, struct rq_flags *rf)
 	struct rq *rq_i;
 	bool need_sync;
 
-	if (!sched_core_enabled(rq))
+	if (likely(!sched_core_enabled(rq)))
 		return __pick_next_task(rq, rf);
 
 	cpu = cpu_of(rq);
@@ -7736,7 +7736,7 @@ void rt_mutex_setprio(struct task_struct *p, struct task_struct *pi_task)
 #if !defined(CONFIG_PREEMPTION) || defined(CONFIG_PREEMPT_DYNAMIC)
 int __sched __cond_resched(void)
 {
-	if (should_resched(0) && !irqs_disabled()) {
+	if (unlikely(should_resched(0) && !irqs_disabled())) {
 		preempt_schedule_common();
 		return 1;
 	}
-- 
2.55.0

diff --git a/fs/readdir.c b/fs/readdir.c
index 76bb1ae3a450..4d9b7a45e72a 100644
--- a/fs/readdir.c
+++ b/fs/readdir.c
@@ -149,7 +149,7 @@ EXPORT_SYMBOL(iterate_dir);
  */
 static int verify_dirent_name(const char *name, int len)
 {
-	if (len <= 0 || len >= PATH_MAX)
+	if (unlikely(len <= 0 || len >= PATH_MAX))
 		return -EIO;
 	if (memchr(name, '/', len))
 		return -EIO;
-- 
2.55.0

diff --git a/include/linux/list.h b/include/linux/list.h
index 09d979976b3b..7dd5b81d9b3e 100644
--- a/include/linux/list.h
+++ b/include/linux/list.h
@@ -151,7 +151,7 @@ static inline bool __list_del_entry_valid(struct list_head *entry)
  * This is only for internal list manipulation where we know
  * the prev/next entries already!
  */
-static inline void __list_add(struct list_head *new,
+static __always_inline void __list_add(struct list_head *new,
 			      struct list_head *prev,
 			      struct list_head *next)
 {
@@ -172,7 +172,7 @@ static inline void __list_add(struct list_head *new,
  * Insert a new entry after the specified head.
  * This is good for implementing stacks.
  */
-static inline void list_add(struct list_head *new, struct list_head *head)
+static __always_inline void list_add(struct list_head *new, struct list_head *head)
 {
 	__list_add(new, head, head->next);
 }
@@ -186,7 +186,7 @@ static inline void list_add(struct list_head *new, struct list_head *head)
  * Insert a new entry before the specified head.
  * This is useful for implementing queues.
  */
-static inline void list_add_tail(struct list_head *new, struct list_head *head)
+static __always_inline void list_add_tail(struct list_head *new, struct list_head *head)
 {
 	__list_add(new, head->prev, head);
 }
@@ -221,7 +221,7 @@ static inline void list_add_tail_release(struct list_head *new,
  * This is only for internal list manipulation where we know
  * the prev/next entries already!
  */
-static inline void __list_del(struct list_head * prev, struct list_head * next)
+static __always_inline void __list_del(struct list_head * prev, struct list_head * next)
 {
 	next->prev = prev;
 	WRITE_ONCE(prev->next, next);
@@ -235,13 +235,13 @@ static inline void __list_del(struct list_head * prev, struct list_head * next)
  * WRITE_ONCE() overhead of a regular list_del_init(). The code that uses this
  * needs to check the node 'prev' pointer instead of calling list_empty().
  */
-static inline void __list_del_clearprev(struct list_head *entry)
+static __always_inline void __list_del_clearprev(struct list_head *entry)
 {
 	__list_del(entry->prev, entry->next);
 	entry->prev = NULL;
 }
 
-static inline void __list_del_entry(struct list_head *entry)
+static __always_inline void __list_del_entry(struct list_head *entry)
 {
 	if (!__list_del_entry_valid(entry))
 		return;
@@ -255,7 +255,7 @@ static inline void __list_del_entry(struct list_head *entry)
  * Note: list_empty() on entry does not return true after this, the entry is
  * in an undefined state.
  */
-static inline void list_del(struct list_head *entry)
+static __always_inline void list_del(struct list_head *entry)
 {
 	__list_del_entry(entry);
 	entry->next = LIST_POISON1;
-- 
2.55.0

diff --git a/drivers/input/evdev.c b/drivers/input/evdev.c
index c7325226cb86..4f18a50ea366 100644
--- a/drivers/input/evdev.c
+++ b/drivers/input/evdev.c
@@ -46,6 +46,7 @@ struct evdev_client {
 	struct fasync_struct *fasync;
 	struct evdev *evdev;
 	struct list_head node;
+	struct rcu_head rcu;
 	enum input_clock_type clk_type;
 	bool revoked;
 	unsigned long *evmasks[EV_CNT];
@@ -368,13 +369,22 @@ static void evdev_attach_client(struct evdev *evdev,
 	spin_unlock(&evdev->client_lock);
 }
 
+static void evdev_reclaim_client(struct rcu_head *rp)
+{
+	struct evdev_client *client = container_of(rp, struct evdev_client, rcu);
+	unsigned int i;
+	for (i = 0; i < EV_CNT; ++i)
+		bitmap_free(client->evmasks[i]);
+	kvfree(client);
+}
+
 static void evdev_detach_client(struct evdev *evdev,
 				struct evdev_client *client)
 {
 	spin_lock(&evdev->client_lock);
 	list_del_rcu(&client->node);
 	spin_unlock(&evdev->client_lock);
-	synchronize_rcu();
+	call_rcu(&client->rcu, evdev_reclaim_client);
 }
 
 static int evdev_open_device(struct evdev *evdev)
@@ -427,7 +437,6 @@ static int evdev_release(struct inode *inode, struct file *file)
 {
 	struct evdev_client *client = file->private_data;
 	struct evdev *evdev = client->evdev;
-	unsigned int i;
 
 	mutex_lock(&evdev->mutex);
 
@@ -439,11 +448,6 @@ static int evdev_release(struct inode *inode, struct file *file)
 
 	evdev_detach_client(evdev, client);
 
-	for (i = 0; i < EV_CNT; ++i)
-		bitmap_free(client->evmasks[i]);
-
-	kvfree(client);
-
 	evdev_close_device(evdev);
 
 	return 0;
@@ -486,7 +490,6 @@ static int evdev_open(struct inode *inode, struct file *file)
 
  err_free_client:
 	evdev_detach_client(evdev, client);
-	kvfree(client);
 	return error;
 }
 
-- 
2.55.0


